### PR TITLE
Session2

### PR DESCRIPTION
**Pull request by gauravcusp**

### CODE REVIEW

File: 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb [hunk 1]
```diff
@@ -16,53 +24,50 @@
...
-      }
-     },
-     "output_type": "execute_result"
-    }
-   ],
    "source": [
-    "from IPython.display import Image\n",
-    "\n",
-    "Image(\"for_loop.png\", embed=True, height=400, width=300)"
+    "While writing a for loop, you must also remember to end the declaration with : and indent the subsidiary code. Let's start with the basic for loop, to address our problem statement."
```
Comment (kbjarkefur) [inline 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb:70]:
```suggestion
    "While writing a for loop, you must also remember to end the declaration with `:` and indent the subsidiary code. Let's start with the basic for loop, to address our problem statement."
```

File: 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb [hunk 2]
```diff
@@ -1,5 +1,13 @@
 {
  "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "d8b75c26-1e53-48bd-86d2-5812947a83fc",
+   "metadata": {},
+   "source": [
+    "[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/googlecolab/colabtools/blob/master/notebooks/colab-github-demo.ipynb)"
```
Comment (kbjarkefur) [inline 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb:8]:
```suggestion
    "[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/worldbank/dec-python-course/blob/main/1-foundations/2-control-flows-and-functions/foundations-s2.ipynb)"
```

This button will now point to this notebook in the `main` branch. Which I think it should, but then it becomes important that we remember to merge this PR before your session

File: 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb [hunk 3]
```diff
@@ -146,17 +156,33 @@
     "    print(country + \"'s GDP in 2020 was \"+ str(gdp_dict[country]) + \" billion US dollars.\")"
```
Comment (kbjarkefur) [inline 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb:156]: I like using `.items()` and a great example of a method/function that returns a typle. If you want, maybe add as an example on how you can do the same thing in two different ways. But fully up to you.
```
for country, gdp in gdp_dict.items():
    print(country + "'s GDP in 2020 was "+ str(gdp) + " billion US dollars.")

File: 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb [hunk 4]
```diff
@@ -510,128 +705,465 @@
...
+   "id": "b4bf2327-69af-4a5a-9170-efa6010cc3fc",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "This is an example of a built in function\n",
+      "The string 'Mississippi' has 11 cahracters\n",
+      "Sorted version of list [7,8,1,4,2] is [1, 2, 4, 7, 8]\n"
```
Comment (kbjarkefur) [inline 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb:956]:
```suggestion
      "Sorted version of list [7,8,1,4,2] is [1,2,4,7,8]\n"
```
Just so no one things the spaces came from `.sorted()`

File: 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb [hunk 5]
```diff
@@ -510,128 +705,465 @@
...
+   "id": "assured-qatar",
+   "metadata": {},
+   "outputs": [],
    "source": [
     "## Function with a parameter with default value\n",
     "\n",
     "def what_day_is_it(day = \"Thursday\"):\n",
+    "    \n",
     "    print(\"Today is {}\".format(day))"
    ]
```
Comment (kbjarkefur) [inline 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb:1162]: Maybe first show what happens if someone use `what_day_is_it()` without parameters before there is a default value. Good learning opportunity to teach participants on how to read error messages when forgetting to pass parameters.

Then after you introduced default value `what_day_is_it(day = "Thursday"):` and show how it works, show what happens if you pass too many parameters

File: 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb [hunk 6]
```diff
@@ -667,11 +1199,13 @@
...
+   "execution_count": 33,
    "id": "greek-gentleman",
    "metadata": {},
    "outputs": [],
    "source": [
+    "##  Creating a parametrized function (This is how your functions would look most of the times when implementing actual use cases)\n",
+    "\n",
     "def query_population(country = 'AFG', series = \"SP.POP.TOTL\", time = 2020):\n",
     "    \"\"\"\n",
     "    Function to retrieve indicator level data from WDI database. \n",
```
Comment (kbjarkefur) [inline 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb:1211]: Not sure what to request instead, but I wonder if this is too advanced. The participants learnt what a list and a float was the day before. While the API topic is very applicable for the audience and promotes WB resources, I wonder if this just risks throwing the participants off. 

Or maybe I there is something I am missing. Perhaps you did not mean for them to fully understand, just to give them an idea of what a real-life function might look like.

File: 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb [hunk 7]
```diff
@@ -726,18 +1268,264 @@
     "print(\"{} million people live in {}. Data as of {}\".format(query_population(\"BRA\", \"SP.POP.TOTL\", 2019), 'BRA', 2019))"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "id": "41fb3ec1-5330-49c4-9675-23a6d7a5effa",
+   "metadata": {},
+   "source": [
+    "### 2.5.3 Local and Global variables (a short detour)"
```
Comment (kbjarkefur) [inline 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb:1276]: I agree that it is important that we mention that a variable created in a function cannot be accessed outside of it. But maybe it is better to just talk about scope and not get into `global`/`local` and be careful to say anything that seems to recommend these beginners to start declaring global variables.

File: 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb (hunk 6 above)
Comment (kbjarkefur) [reply 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb:1211]: As you keep using this function I think you should keep it in. But maybe make sure to make it clear that they are not expected to understand the details of an API. But the code should make some sense. What they really should be able to figure out is how to use it.

File: 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb [hunk 8]
```diff
@@ -726,18 +1268,264 @@
...
+   "cell_type": "code",
+   "execution_count": 148,
+   "id": "bc964d53-2076-47e9-b304-a258dba6b394",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "### YOUR CODE HERE\n",
+    "\n",
+    "## HINT: You'll have to make two requests to get data for two variables GDP and Population. Series code for GDP is \"NY.GDP.MKTP.CD\" and for Population is \"NY.GDP.MKTP.CD\". \n",
+    "## You can take help from query_population function and substitute the series code with the one provided above. Index the request and get the values returned for that country and time period. \n",
```
Comment (kbjarkefur) [inline 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb:1505]:
```suggestion
    "## HINT: You'll have to make two requests to get data for two variables GDP and Population. \n"
    "## Series code for GDP is \"NY.GDP.MKTP.CD\" and for Population is \"NY.GDP.MKTP.CD\". \n",
    "## You can take help from query_population function and substitute the series code with \n"
    "##  the one provided above. Index the request and get the values returned for that country and time period. \n",
```
Markdown wraps text but not comments. Breaking this up to make it readable without scrolling

File: 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb (hunk 5 above)
Comment (gauravcusp) [reply 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb:1162]: Good point!

File: 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb (hunk 6 above)
Comment (gauravcusp) [reply 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb:1211]: I'll be sure to mention not to pay much attention with API but just string concatenation for now.

File: 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb (hunk 7 above)
Comment (gauravcusp) [reply 1-foundations/2-control-flows-and-functions/foundations-s2.ipynb:1276]: Fair point!

**Comment by gauravcusp** [comment]
Hey @kbjarkefur I've addressed your comments. Let me know if you have any other concerns. Otherwise I'll merge it with main,

**Comment by kbjarkefur** [comment]
Did you see this: https://github.com/worldbank/dec-python-course/pull/10

Sorry for creating a new branch, but since most of the session was already merged to `main` it was not straightforward how to review the full session in this PR.

**Pull request merged**